cmd/katapult/output.go

```diff
@@ -0,0 +1,213 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"sort"
+	"strings"
+	"text/template"
+
+	"github.com/olekukonko/tablewriter"
+	"github.com/spf13/cobra"
+	"gopkg.in/yaml.v2"
+)
+
+var outputFlag, templateFlag string
+
+// Output is used to define the interface of outputs.
+type Output interface {
+	// JSON is used to return a string of the JSON output.
+	JSON() (json.RawMessage, error)
+
+	// YAML is used to return a string of the YAML output.
+	YAML() (string, error)
+
+	// Text is used to render a template. If string is blank, uses the default.
+	Text(template string) (string, error)
+}
+
+// Used to implement Output for a variety of use cases.
+type genericOutput struct {
+	item interface{}
+	tpl  string
+}
+
+// JSON is used to return a string of the JSON output.
+func (g genericOutput) JSON() (json.RawMessage, error) {
+	b, err := json.Marshal(g.item)
+	if err != nil {
+		return nil, err
+	}
+	return b, nil
+}
+
+// YAML is used to return a string of the YAML output.
+func (g genericOutput) YAML() (string, error) {
+	b, err := yaml.Marshal(g.item)
+	if err != nil {
+		return "", err
+	}
+	return string(b), nil
+}
+
+// Used to render a table.
+func table(columns []string, rows [][]interface{}) string {
```

Comment:
The output tables look really nice, but I'm not sure they're the most useful way to present tabular data in a console, where it would be quite common to further process the output with `grep`, `awk` and similar tools. And all the extra characters to render a visual table around the data itself can make that a bit more difficult.

Hence, what do you think about simplifying the output a bit to be just structured text?

For example, the list of VMs looks like this right now:

```
+-------------------------+----------------------------------------------+
|          NAME           |                     FQDN                     |
+-------------------------+----------------------------------------------+
| popular-silent-bluebell | popular-silent-bluebell.jimeh.katapult.cloud |
| powerful-loud-balloon   | powerful-loud-balloon.jimeh.katapult.cloud   |
| worried-silent-panda    | worried-silent-panda.jimeh.katapult.cloud    |
+-------------------------+----------------------------------------------+
```

Instead I'd like to propose we render this list something like this:

```
NAME                     FQDN
popular-silent-bluebell  popular-silent-bluebell.jimeh.katapult.cloud
powerful-loud-balloon    powerful-loud-balloon.jimeh.katapult.cloud
worried-silent-panda     worried-silent-panda.jimeh.katapult.cloud
```

It makes it easy to extract specific pieces of data, like the FQDN for example, by first removing the first line with `tail -n +1` and then pulling out the second column with `awk '{ print $2 }'`, the result of which you could then loop over and give to `katapult vm stop` for example.

Reply:
No, that seems quite reasonable.